lib/cog/command/pipeline/executor.ex

```diff
@@ -198,169 +181,107 @@ defmodule Cog.Command.Pipeline.Executor do
+        # TODO: what is this error case?
+        Helpers.send_error(msg, state.request, state.mq_conn)
         fail_pipeline(state, :binding_error, "Error preparing to execute command pipeline '#{state.request["text"]}': #{msg}")
-      bound ->
-        {current_bound, bound_scope} = collect_bound(bound)
-        {:next_state, :maybe_expand, %{state | current_bound: current_bound, scope: bound_scope}, 0}
-    end
-  end
-
-  @doc """
-  `expand` -> {:stop, :shutdown} | {:next_state, :get_options}
-
-  If a command begins with 'user' or 'site' it is treated as an alias and expanded.
-  """
-  def maybe_expand(:timeout, %__MODULE__{current: current_bound}=state) do
-    case AliasExpander.expand(current_bound.command, state.user) do
-      {:ok, :not_an_alias} ->
-        {:next_state, :get_options, state, 0}
-      {:ok, invocations} ->
-        [new_current|rest] = invocations
-        {:next_state, :bind, %{state | current: new_current, remaining: rest ++ state.remaining}, 0}
-      {:error, msg} ->
-        Helpers.send_reply(msg, state.request, state.mq_conn)
-        fail_pipeline(state, :alias_expansion_error, "Error expanding alias '#{current_bound.command}': #{msg}")
-    end
-  end
-
-  @doc """
-  `get_options` -> {:stop, :shutdown} | {:next_state, :check_permission}
-
-  Runs the option interpreter on the current bound invocation.
-  """
-  def get_options(:timeout, %__MODULE__{current_bound: current_bound}=state) do
-    case interpret_options(current_bound) do
-      {:ok, current_bound} ->
-        {:next_state, :maybe_enforce, %{state | current_bound: current_bound}, 0}
-      {:not_found, current_bound} ->
-        Helpers.send_idk(state.request, current_bound.command, state.mq_conn)
-        fail_pipeline(state, :option_interpreter_error, "Command '#{current_bound.command}' not found")
       error ->
         {:error, msg} = error
         Helpers.send_error(msg, state.request, state.mq_conn)
         fail_pipeline(state, :option_interpreter_error, "Error parsing options: #{inspect error}")
     end
   end
+  def plan_next_invocation(:timeout, %__MODULE__{invocations: [], output: [], destinations: destinations}=state) do
+    message = "Pipeline executed successfully, but no output was returned"
+    adapter = state.request["adapter"]
+    Enum.each(destinations, &publish_response(response_fn(message, adapter), &1, state))
+    {:stop, :shutdown, %{state | output: []}}
+  end
+  def plan_next_invocation(:timeout, %__MODULE__{invocations: [], output: output, destinations: destinations}=state) do
+    # No more invocations, but we've got data; run it through
+    # templating and send it out to all the destinations
+    command = state.current_plan.command
+    adapter = state.request["adapter"]
 
-  @doc """
-  `maybe_enforce` -> {:next_state, :maybe_collect_command} | {:next_state, :check_permission}
+    message = Enum.map_join(output, "\n", fn({data,template}) ->
+      render_template(command, adapter, template, data)
```

Comment:
Let's say there's a pipeline that looks conceptually like this:
```
ec2-list --region=-us-east-1 | ec2-terminate --region=us-east-1 $instance_id
```

The `ec2-list` stage returns:

```
[
  { "instance_id": "i-al8f3a" },
  { "instance_id": "i-1ab34b" },
  { "instance_id": "i-9b90ea" }
]
```

The first two executions of `ec2-terminate` get run normally, but know that they are not processing the final input. In this case, rather than returning a real result to the executor, they store their result in a state store that the final execution can access and return an empty result which will not get rendered.

For demonstration purposes let's assume the following:

```
Execution: ec2-terminate --region=us-east-1 i-al8f3a
Last: false
Stored Result: { "status": "terminated", "instance_id": "i-al8f3a" }
Returned Result: []
```

```
Execution: ec2-terminate --region=us-east-1 i-1ab34b
Last: false
Stored Result: { "status": "terminated", "instance_id": "i-1ab34b" }
Returned Result: []
```

The final execution is called exactly the same way, but has an additional piece of information in the environment to tell it that it's the last item in the stage.

```
Execution: ec2-terminate --region=us-east-1 i-1ab34b
Last: true
```

The command would perform its normal work, generating the following result:

```
Result: { "status": "terminated", "instance_id": "i-1ab34b" }
```

It could then retrieve the results from the previous executions in the stage and could return the following, which could get rendered by a single template:

```
[
  { "status": "terminated", "instance_id": "i-al8f3a", "region": "us-east-1" },
  { "status": "terminated", "instance_id": "i-1ab34b", "region": "us-east-1" },
  { "status": "termianted", "instance_id": "i-9b90ea", "region": "us-east-1" }
]
```